### PR TITLE
[Issue 3063] Add BeaconBlockSummary interface

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Lay some groundwork for starting up from an initial state without requiring a matching block by adding a  `BeaconBlockSummary` interface.  This interface makes it easier to work with state and block header information without requiring full blocks.  

Summary:
* Add `BeaconBlockSummary` interface
* Update `Store` state regeneration tasks to return the new summary type
* Update `AnchorPoint` to hold a `BeaconBlockSummary` rather than requiring a full block
* Update `ChainHead` to hold a `BeaconBlockSummary`, so that we can handle chain heads that do not have a full block available
* Update `CheckpointState` to hold a `BeaconBlockSummary` type rather than a full block

## Fixed Issue(s)
Part of #3063 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/blocks/BeaconBlockHeader.java [hunk 1]
```diff
@@ -87,6 +94,43 @@ public BeaconBlockHeader() {
...
+    return latestHeader;
+  }
+
+  public static BeaconBlockHeader fromBlock(final BeaconBlock block) {
+    return new BeaconBlockHeader(
+        block.getSlot(),
+        block.getProposerIndex(),
+        block.getParentRoot(),
+        block.getStateRoot(),
+        block.getBody().hash_tree_root());
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/blocks/BeaconBlockHeader.java]: nit: I think you can just use:
```suggestion
        block.getBodyRoot());
```

**Review by ajsutton — COMMENTED** [review]
Looks like a really promising approach. Will be interesting to see if memory usage is reduced by not holding so many block bodies all the time.

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/AnchorPoint.java [hunk 2]
```diff
@@ -30,33 +34,36 @@
...
-    this.state = state;
     this.isGenesis = checkpoint.getEpoch().equals(UInt64.valueOf(Constants.GENESIS_EPOCH));
-    this.blockAndState = new SignedBlockAndState(block, state);
+  }
+
+  public static AnchorPoint create(
+      Checkpoint checkpoint, BeaconState state, Optional<SignedBeaconBlock> block) {
+    final BeaconBlockSummary blockSummary =
+        block
+            .map(BeaconBlockSummary.class::cast)
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/AnchorPoint.java]: This is potentially a bit heavy handed as it removes all compile time checking.  Can we get away with:
```suggestion
            .<BeaconBlockSummary>map(a -> a)
```

It doesn't read quite as well, but will give compile errors if `block` ever winds up not actually being a `BeaconBlockSummary`.

**Review by ajsutton — APPROVED** [review]
LGTM assuming the conflicts resolve ok.

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/AnchorPoint.java (hunk 2 above)
Comment (mbaxter) [reply ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/AnchorPoint.java]: good call 👍

**Pull request merged**